File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No114FlattenBinaryTreeToLinkedList.java
package com.hazza.LeetCodeByJava.SegmentTwo;

import com.hazza.LeetCodeByJava.datastructure.TreeNode;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/flatten-binary-tree-to-linked-list/description/
 * User: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 18-1-26
 * Time: 2:06 PM
 */
public class No114FlattenBinaryTreeToLinkedList {

    // solution 1
    public void flatten(TreeNode root) {
        flatten(root, null);
    }

    private TreeNode flatten(TreeNode node, TreeNode prev) {
        if (node == null) return null;
        prev = flatten(node.right, prev);
        prev = flatten(node.left, prev);
        node.right = prev;
        node.left = null;

        return node;
    }

    // solution 2
    public void flatten1(TreeNode root) {
        if (root == null) return;
        TreeNode cur = root;
        while (cur != null) {
            if (cur.left != null) {
                TreeNode prev = cur.left;
                // find the previous node
                while (prev.right != null) prev = prev.right;
                prev.right = cur.right;
                cur.right = cur.left;
                cur.left = null;
            }
            cur = cur.right;
        }
    }

    // solution 3
    public void flatten2(TreeNode root) {
        if (root == null) return;
        Deque<TreeNode> stack = new ArrayDeque<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode cur = stack.pop();
            if (cur.right != null) stack.push(cur.right);
            if (cur.left != null) stack.push(cur.left);
            if (!stack.isEmpty()) cur.right = stack.peek();
            cur.left = null;
        }
    }
}

File: src/main/python/leetcode_by_python/tree/No114_Flatten_Binary_Tree_to_Linked_List.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-10-31
from src.main.python.leetcode_by_python.data_structure.TreeNode import TreeNode


class Solution:
    # Solution 1
    def flatten1(self, root: TreeNode) -> None:
        self.prev = None

        def dfs(node):
            if not node:
                return None
            dfs(node.right)
            dfs(node.left)
            node.right = self.prev
            node.left = None
            self.prev = node

        dfs(root)

    # Solution 2
    def flatten2(self, root: TreeNode) -> None:
        if not root:
            return None
        cur = root
        while cur:
            if cur.left:
                prev = cur.left
                while prev.right:
                    prev = prev.right
                prev.right = cur.right
                cur.right = cur.left
                cur.left = None
            cur = cur.right

    # Solution 3
    def flatten3(self, root: TreeNode) -> None:
        if not root:
            return None
        stack = [root]
        while stack:
            cur = stack.pop()
            if cur.right:
                stack.append(cur.right)
            if cur.left:
                stack.append(cur.left)
            if stack:
                cur.right = stack[-1]
            cur.left = None
